src/encoded/static/components/report.js

```diff
@@ -282,53 +283,190 @@ ColumnHeader.propTypes = {
 };
 
 
-class ColumnSelector extends React.Component {
+class ColumnSelectorControls extends React.Component {
     constructor() {
         super();
 
         // Set initial React state.
         this.state = {
-            open: false,
+            selectedSort: 'default',
         };
 
-        // Bind this to non-React methods.
-        this.toggle = this.toggle.bind(this);
+        // Bind `this` to non-React methods.
+        this.handleSortChange = this.handleSortChange.bind(this);
+    }
+
+    handleSortChange(e) {
+        // Called when the user changes the sorting option. Sets a component state so that the
+        // controlled <select> component renders properly. It then calls the parent's callback to
+        // react to the new sorting option.
+        this.setState({ selectedSort: e.target.value });
+        this.props.handleSortChange(e.target.value);
+    }
+
+    render() {
+        const { handleSelectAll, handleSelectOne, firstColumnTitle } = this.props;
+
+        return (
+            <div className="column-selector__controls">
+                <div className="column-selector__utility-buttons">
+                    <button onClick={handleSelectAll} className="btn btn-info">Select all</button>
+                    <button onClick={handleSelectOne} className="btn btn-info">Select {firstColumnTitle} only</button>
+                </div>
+                <div className="column-selector__sort-selector">
+                    <select className="form-control--select" value={this.state.selectedSort} onChange={this.handleSortChange}>
+                        <option value="default">Default sort</option>
+                        <option value="alpha">Alphabetical sort</option>
+                    </select>
+                </div>
+            </div>
+        );
+    }
+}
+
+ColumnSelectorControls.propTypes = {
+    handleSelectAll: PropTypes.func.isRequired, // Callback when Select All button clicked
+    handleSelectOne: PropTypes.func.isRequired, // Callback when SelectOne button clicked
+    handleSortChange: PropTypes.func.isRequired, // Callback when sorting option changed
+    firstColumnTitle: PropTypes.string.isRequired, // Title of first column
+};
+
+
+// Displays a modal dialog with every possible column for the type of object being displayed.
+// This lets you choose which columns you want to appear in the report.
+class ColumnSelector extends React.Component {
+    constructor(props) {
+        super(props);
+
+        // Set the initial React states.
+        this.state = {
+            columns: props.columns, // Make (effectively) a local mutatable copy of the columns
```

Comment:
This doesn't make a mutable copy; it just adds a reference to the object that came in via props. We don't want to mutate an object that's controlled by another component, so we need to be careful to construct a new object whenever we make changes to this.state.columns.

Reply:
Is there a reference to this? I mean, the initial state is definitely a reference to the original props object, but setState _seems to_ use a copy, as I see no changes to the parent columns object when changes are made to this.state.columns in this component. The React docs don’t quite touch the topic, only saying to be careful about using props to set initial state, and I think this modal case falls under the OK-to-do case since it’s a modal with its own version of columns. I could see the argument that it’s just risky even if not documented.

Reply:
It may work in practice if React is giving us a copy, but since the React docs say not to do it, I would not assume that it will always continue to work.

Reply:
As long as you're going to change the code below that mutates things, it should be okay.

Reply:
React docs don’t say not to do it, but their example uses a simple value instead of an object, so mutating props wasn’t their concern. Their “Beware” note doesn’t apply to this modal, since if the props change, the modal would be unmounted. https://reactjs.org/docs/react-component.html#constructor